### PR TITLE
mongodb_user: fix exception handling in user edit/update

### LINKED ISSUE #2575: mongodb_user throws exception instead of failing properly
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

mongodb_user
##### ANSIBLE VERSION

ansible 2.2.0 (devel 7ecc75b555)
##### OS / ENVIRONMENT

N/A
##### SUMMARY

When mongodb_user tries to add an user, first it checks for that user. During the check, an exception can occur, which is not caught at all.
##### STEPS TO REPRODUCE

mongodb_user: database=admin name=admin password=a_password roles=userAdminAnyDatabase

An exception occurred during task execution. To see the full traceback, use -vvv. The error was: pymongo.errors.OperationFailure: database error: not authorized for query on admin.system.users
##### EXPECTED RESULTS

The module would fail gracefully
##### ACTUAL RESULTS

The module throws an exception

```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: pymongo.errors.OperationFailure: database error: not authorized for query on admin.system.users
fatal: [jsm-devops-openstack-controller-0]: FAILED! => {"changed": false, "failed": true, "failed_when_result": true, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_QnOUu8/ansible_module_mongodb_user.py\", line 376, in <module>\n    main()\n  File \"/tmp/ansible_QnOUu8/ansible_module_mongodb_user.py\", line 347, in main\n    uinfo = user_find(client, user, db_name)\n  File \"/tmp/ansible_QnOUu8/ansible_module_mongodb_user.py\", line 188, in user_find\n    for mongo_user in client[\"admin\"].system.users.find():\n  File \"/usr/lib/python2.7/dist-packages/pymongo/cursor.py\", line 1097, in next\n    if len(self.__data) or self._refresh():\n  File \"/usr/lib/python2.7/dist-packages/pymongo/cursor.py\", line 1019, in _refresh\n    self.__read_concern))\n  File \"/usr/lib/python2.7/dist-packages/pymongo/cursor.py\", line 903, in __send_message\n    codec_options=self.__codec_options)\n  File \"/usr/lib/python2.7/dist-packages/pymongo/helpers.py\", line 128, in _unpack_response\n    error_object)\npymongo.errors.OperationFailure: database error: not authorized for query on admin.system.users\n", "module_stdout": "", "msg": "MODULE FAILURE", "parsed": false}
```

### PR DESCRIPTION
**Pull request by gyurco**
##### ISSUE TYPE
- Bugfix Pull Request
##### COMPONENT NAME

mongodb_user
##### ANSIBLE VERSION

ansible 2.2.0 (devel 7ecc75b555)
##### SUMMARY

Properly guard all pymongo operations with a try...except block

Fixes: #2575

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @gyurco. To the current maintainers, @lujeni please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

File: database/misc/mongodb_user.py [hunk 1]
```diff
@@ -344,16 +342,16 @@ def main():
...
+        try:
+            uinfo = user_find(client, user, db_name)
+            if update_password != 'always' and uinfo:
+                password = None
+                if not check_if_roles_changed(uinfo, roles, db_name):
+                    module.exit_json(changed=False, user=user)
 
-        if module.check_mode:
-            module.exit_json(changed=True, user=user)
+            if module.check_mode:
```
Comment (Lujeni) [inline database/misc/mongodb_user.py:352]: the indentation seems weird (tab?), can you check pls

**Comment by Lujeni** [comment]
Hi,

Thanks for your pull request/fix.
- [ ] Can you remove the first commit about the ssl bug pls (duplicate with https://github.com/ansible/ansible-modules-extras/pull/2573)
- [ ] check the indentation comment

Regard

File: database/misc/mongodb_user.py (hunk 1 above)
Comment (gyurco) [inline database/misc/mongodb_user.py:352]: Checked with a hex editor, they're spaces (0x20)

**Comment by gyurco** [comment]
I'm not sure how to remove the first commit, since this has built on it (however logically they're not connected). Maybe if the previous one is merged (since thanks to you, it got the shipit), it'll disappear from here?

**Comment by gyurco** [comment]
Hi @Lujeni,

It is now one commit only.

Br,
György

**Comment by Lujeni** [comment]
> It is now one commit only

Thanks

`shipit`

**Comment by gregdek** [comment]
Thanks again to @gyurco for this PR, and thanks @lujeni for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Pull request merged**